### PR TITLE
Futher storage isolation and removal of the tools interface.

### PR DESCRIPTION
**Pull request by timothysc**
This completes the removal of the tools.interface and adds mechanics to allow for multiple back-ends via a Config interface. 

@wojtek-t & @karlkfi | @jdef for review, as it touches their areas.

### CODE REVIEW

**Comment by wojtek-t** [comment]
Will take a look tomorrow.

**Comment by karlkfi** [comment]
Why do we have a forked go-etcd? 
Is that what's allowing us to remove our client interface? 
Trading a util package and interface for a fork to maintain doesn't seem like an improvement...

**Comment by timothysc** [comment]
@karlkfi - The fork has nothing to do with changes to your code, they are shuffling to isolate the dependency which existed in the startup of the apiserver. 

The modifications to the mesos code are simply to remove the tools dependency.

File: contrib/mesos/pkg/election/etcd_master_test.go [hunk 1]
```diff
@@ -57,77 +54,21 @@ func TestEtcdMasterNoOther(t *testing.T) {
...
 	if result.Type != watch.Modified || result.Object.(Master) != "baz" {
 		t.Errorf("unexpected event: %#v", result)
 	}
 	w.Stop()
-
-	expectedCalls := []string{"get", "create", "get"}
-	if !reflect.DeepEqual(client.calls, expectedCalls) {
-		t.Errorf("unexpected calls: %#v", client.calls)
-	}
+	w_ldr.Stop()
```
Comment (karlkfi) [inline contrib/mesos/pkg/election/etcd_master_test.go]: do we not need to show that the calls happened in the expected order?

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -283,36 +278,21 @@ func (s *APIServer) verifyClusterIPFlags() {
...
-			Dial: forked.Dial,
-			TLSClientConfig: &tls.Config{
-				InsecureSkipVerify: true,
-			},
-			MaxIdleConnsPerHost: 500,
-		}
-		etcdClient.SetTransport(transport)
-		client = etcdClient
-	}
+	var storagCfg etcdstorage.EtcdConfig
```
Comment (wojtek-t) [inline cmd/kube-apiserver/app/server.go]: s/storagCfg/storageCfg/

or even storageConfig

File: pkg/storage/interfaces.go [hunk 3]
```diff
@@ -149,3 +149,13 @@ type Interface interface {
 	// Codec provides access to the underlying codec being used by the implementation.
 	Codec() runtime.Codec
 }
+
+// Config interface allows storage teirs to generate the proper storage.interface
```
Comment (wojtek-t) [inline pkg/storage/interfaces.go]: s/teirs/tiers/ ?

**Comment by wojtek-t** [comment]
The changes outside mesos code lgtm (just very minor comments).

File: contrib/mesos/pkg/election/etcd_master_test.go (hunk 1 above)
Comment (timothysc) [inline contrib/mesos/pkg/election/etcd_master_test.go]: I would argue no, b/c the goal of the test is to check that (X) is not the leader.  

Operational ordering imho was a false test that wrapped the tool.interface.

File: cmd/kube-apiserver/app/server.go (hunk 2 above)
Comment (timothysc) [inline cmd/kube-apiserver/app/server.go]: done.

File: pkg/storage/interfaces.go (hunk 3 above)
Comment (timothysc) [inline pkg/storage/interfaces.go]: done.

**Comment by wojtek-t** [comment]
LGTM from mi side (I'm waiting for @karlkfi for confirmation if he is also fine with it)

**Comment by jdef** [comment]
lgtm

**Comment by timothysc** [comment]
@wojtek-t PTAL for lgtm apply.

**Comment by wojtek-t** [comment]
Generated docs are out of date. Please run hack/update-generated-docs.sh
!!! Error in hack/verify-generated-docs.sh:28
  '"${KUBE_ROOT}/hack/after-build/verify-generated-docs.sh" "$@"' exited with status 1

@timothysc - can you please fix it?

**Comment by timothysc** [comment]
@wojtek-t done.

**Comment by wojtek-t** [comment]
LGTM

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -234,7 +230,6 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.AdmissionControl, "admission-control", s.AdmissionControl, "Ordered list of plug-ins to do admission control of resources into cluster. Comma-delimited list of: "+strings.Join(admission.GetPlugins(), ", "))
 	fs.StringVar(&s.AdmissionControlConfigFile, "admission-control-config-file", s.AdmissionControlConfigFile, "File with admission control configuration.")
 	fs.StringSliceVar(&s.EtcdServerList, "etcd-servers", s.EtcdServerList, "List of etcd servers to watch (http://ip:port), comma separated. Mutually exclusive with -etcd-config")
-	fs.StringVar(&s.EtcdConfigFile, "etcd-config", s.EtcdConfigFile, "The config file for the etcd client. Mutually exclusive with -etcd-servers.")
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: This is a breaking change... we now have no way to run against a secured etcd. I know the client lib no longer supports the config file, but we have to (even if it means deserializing the file and building the etcd client config ourselves)

**Pull request merged**